### ISSUE TITLE
socketConnect event after connection timeout

### ISSUE THREAD

**Issue by hellivan** [tediousjs/tedious#683]
Node version: 8.7.0 
Tedious version: 2.1.5

When connecting to a database, it is possible that a 'socketConnect' event is emitted even after the connection had already timed out. As a result we get a unhandled exception that crashes the whole node application:

```
Error: No event 'socketConnect' in state 'Final'
     at Connection.dispatchEvent ([...]/mssql/node_modules/tedious/lib/connection.js:859:28)
     at Connection.socketConnect ([...]/mssql/node_modules/tedious/lib/connection.js:883:19)
     at [...]/mssql/node_modules/tedious/lib/connection.js:755:16
     at Socket.onConnect ([...]/mssql/node_modules/tedious/lib/connector.js:176:9)
     at emitNone (events.js:67:13)
     at Socket.emit (events.js:166:7)
     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1068:10)
```

The error can be reliably reproduced be setting the connectTimout to 1 ms.

A possible solution for this problem would be to check the connection state (if it already transitioned to state FINAL) just before calling socketConnect.

**Comment by hellivan** [comment]
#682 already provides a fix for this issue.

**Comment by arthurschreiber** [comment]
A fix for this was just released as part of `tedious@2.2.4` (currently still tagged as `next`).

You can give it a try by doing `npm install tedious@2.2.4` or `npm install tedious@next`. Please let us know if you run into any issues.

If we don't hear back from you or anyone else regarding any issues with `tedious@2.2.4`, it will be tagged as `latest` soon! 👍

**Comment by arthurschreiber** [comment]
As we haven't heard back so far, I tagged `2.2.4` as `latest`:

```
$ npm dist-tag add tedious@2.2.4 latest
+latest: tedious@2.2.4
```
